include/logging/log_msg2.h

```diff
@@ -34,6 +35,12 @@ typedef uint64_t log_timestamp_t;
 typedef uint32_t log_timestamp_t;
 #endif
 
+#if !defined(CONFIG_LOG) && \
+	!GCC_VERSION_AT_LEAST(4, 9, 0) && \
+	!CLANG_VERSION_AT_LEAST(3, 8, 0)
```

Comment:
Nitpick: abstraction-wise, might be better to have toolchain.h define "TOOLCHAIN_HAS_AUTO_TYPE" or somesuch instead of putting the logic here in logging.

Reply:
It also guards against using `_Pragma()` within functions, so not exactly about auto type only.

Reply:
i'm also for moving flags to toolchain.h.
`TOOLCHAIN_HAS_AUTO_TYPE`, `TOOLCHAIN_HAS_C11_GENERIC`, maybe `TOOLCHAIN_HAS_DIAGNOSTIC_PRAGMAS` (seems that this was introduced in 4.6).

It's already third place i'm aware of where we do that (one in PR https://github.com/zephyrproject-rtos/zephyr/blob/9705d17e11cd119269d3e7ddba505af8dac0d56b/include/event_manager/event_manager_priv.h#L33).